### PR TITLE
feature: add supernode disk gc feature

### PR DESCRIPTION
**Pull request by Starnop**
Signed-off-by: Starnop <starnop@163.com>

### Ⅰ. Describe what this PR did
With supernode downloading more and more tasks, and it will take up more and more disk space. So we need  a gc mechanism to clean up the disk space occupied by the supenrode when necessary. And that's what this PR does.

### Ⅱ. Does this pull request fix one issue?

#867

### CODE REVIEW

File: supernode/daemon/mgr/gc/gc_manager.go [hunk 1]
```diff
@@ -68,6 +68,17 @@ func (gcm *Manager) StartGC(ctx context.Context) {
...
 	}()
+
+	go func() {
+		// delay to execute GC after gcm.initialDelay
+		time.Sleep(gcm.cfg.GCInitialDelay)
+
+		// execute the GC by fixed delay
+		ticker := time.NewTicker(gcm.cfg.GCDiskInterval)
+		for range ticker.C {
+			go gcm.gcDisk(ctx)
```
[the remark this answers is not in the document]
Comment (Starnop) [reply supernode/daemon/mgr/gc/gc_manager.go]: Done.

File: supernode/daemon/mgr/cdn/cdn_gc.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+func (cm *Manager) GetGCTaskIDs(ctx context.Context, taskMgr mgr.TaskMgr) ([]string, error) {
+	var gcTaskIDs []string
+
+	freeDisk, err := cm.cacheStore.GetAvailSpace(ctx, getHomeRawFunc())
+	if err != nil {
+		return nil, err
+	}
+	if freeDisk > cm.cfg.YoungGCThreshold {
+		logrus.Warnf("no need to gc with free disk(%d) and youngGCThreshold(%d)",
```
Comment (lowzj) [inline supernode/daemon/mgr/cdn/cdn_gc.go]: It's inappropriate to output a warn level log here.

File: supernode/daemon/mgr/gc/gc_disk.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+		}
+
+		if err := gcm.cdnMgr.Delete(ctx, taskID, true); err != nil {
+			util.ReleaseLock(taskID, false)
+			logrus.Errorf("gc disk: failed to delete disk files with taskID(%s): %v", taskID, err)
+			continue
+		}
+		util.ReleaseLock(taskID, false)
+		count++
+
```
Comment (lowzj) [inline supernode/daemon/mgr/gc/gc_disk.go]: delete blank line

File: supernode/daemon/mgr/gc/gc_disk.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		return
+	}
+
+	if len(gcTaskIDs) == 0 {
+		logrus.Warnf("gc disk: get empty gcTasks")
+		return
+	}
+
+	logrus.Debugf("gc disk: success to get gcTaskIDs(%v)", gcTaskIDs)
+	go gcm.deleteTaskDisk(ctx, gcTaskIDs)
```
Comment (lowzj) [inline supernode/daemon/mgr/gc/gc_disk.go]: * don't run it in a different routine
* output the gc result at the end, it's useful for troubleshooting

Comment (Starnop) [reply supernode/daemon/mgr/gc/gc_disk.go]: Done. 

And I output the gc result at the function `deleteTaskDisk `.

File: supernode/daemon/mgr/cdn/cdn_gc.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+		tasks := v.([]string)
+		tasks = append(tasks, metaData.TaskID)
+		intervalTasks.Put(info.Size, tasks)
+		return nil
+	}
+
+	v, found := gapTasks.Get(gap)
+	if !found {
+		v = make([]string, 0)
+
```
Comment (lowzj) [inline supernode/daemon/mgr/cdn/cdn_gc.go]: remove the blank line

File: supernode/daemon/mgr/cdn/cdn_gc.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	walkTaskIDs := make(map[string]bool, 0)
+	walkFn := func(path string, info os.FileInfo, err error) error {
+		logrus.Debugf("start to walk path(%s)", path)
+
+		if err != nil {
+			logrus.Errorf("failed to access path(%s): %v", path, err)
+			return err
+		}
+		if info.IsDir() {
+			logrus.Warnf("file is not regular file(%s)", path)
```
Comment (lowzj) [inline supernode/daemon/mgr/cdn/cdn_gc.go]: delete this log. There're many sub-directories in download home.

Comment (Starnop) [reply supernode/daemon/mgr/cdn/cdn_gc.go]: Done.

**Comment by KeepCaim** [comment]
When is the deadline for this PR?

**Comment by yeya24** [comment]
> When is the deadline for this PR?

What do you mean of the deadline?

**Pull request merged**